test/cpp/end2end/xds/xds_pick_first_end2end_test.cc

```diff
@@ -0,0 +1,170 @@
+#include "src/core/ext/filters/client_channel/resolver/fake/fake_resolver.h"
+#include "src/core/lib/address_utils/sockaddr_utils.h"
+#include "src/core/lib/config/config_vars.h"
+#include "src/core/lib/gprpp/env.h"
+#include "src/proto/grpc/testing/xds/v3/cluster.grpc.pb.h"
+#include "src/proto/grpc/testing/xds/v3/pick_first.pb.h"
+#include "test/core/util/scoped_env_var.h"
+#include "test/core/util/test_config.h"
+#include "test/cpp/end2end/connection_attempt_injector.h"
+#include "test/cpp/end2end/xds/xds_end2end_test_lib.h"
+
+namespace grpc {
+namespace testing {
+namespace {
+
+using ::envoy::extensions::load_balancing_policies::pick_first::v3::PickFirst;
+
+class PickFirstTest : public XdsEnd2endTest {
+ protected:
+  absl::variant<Status, size_t> WaitForAnyBackendHit(size_t start, size_t end) {
+    Status status;
+    size_t index = 3;
+    SendRpcsUntil(DEBUG_LOCATION, [&](const RpcResult& result) -> bool {
+      if (!result.status.ok()) {
+        status = result.status;
+        return false;
+      }
+      for (size_t i = start; i < end; ++i) {
+        if (backends_[i]->backend_service()->request_count() > 0) {
+          backends_[i]->backend_service()->ResetCounters();
+          index = i;
+          return false;
+        }
+      }
+      return true;
+    });
+    if (!status.ok()) {
+      return status;
+    }
+    return index;
+  }
+
+  void ConfigureCluster(size_t iteration, bool shuffle,
+                        size_t first_endpoint_index,
+                        size_t stop_endpoint_index) {
+    std::string edsService = absl::StrCat("edsservice", iteration);
+    // First 3 endpoints go to a service that will have regular pick_first LB
+    balancer_->ads_service()->SetEdsResource(BuildEdsResource(
+        EdsResourceArgs(
+            {{"locality0", CreateEndpointsForBackends(first_endpoint_index,
+                                                      stop_endpoint_index)}}),
+        edsService));
+    PickFirst pick_first;
+    pick_first.set_shuffle_address_list(shuffle);
+    auto cluster = default_cluster_;
+    cluster.clear_lb_policy();
+    cluster.clear_load_balancing_policy();
+    cluster.mutable_eds_cluster_config()->set_service_name(edsService);
+    cluster.mutable_load_balancing_policy()
+        ->add_policies()
+        ->mutable_typed_extension_config()
+        ->mutable_typed_config()
+        ->PackFrom(pick_first);
+
+    // Set ordered cluster configuration
+    balancer_->ads_service()->SetCdsResource(cluster);
+  }
+
+  grpc_core::RefCountedPtr<grpc_core::FakeResolverResponseGenerator>
+      logical_dns_cluster_resolver_response_generator_;
+};
+
+// Run both with and without load reporting, just for test coverage.
+INSTANTIATE_TEST_SUITE_P(
+    XdsTest, PickFirstTest,
+    ::testing::Values(XdsTestType(), XdsTestType().set_enable_load_reporting()),
+    &XdsTestType::Name);
+
+TEST_P(PickFirstTest, PickFirstConfigurationIsPropagated) {
```

Comment:
I think this test is way more complicated than it needs to be.  I think that what we actually want here is something similar to the logic that we have in the [existing unit test](https://github.com/grpc/grpc/blob/21f2eba14389ddb8346dfd16ac2c306037d3c7e9/test/core/client_channel/lb_policy/pick_first_test.cc#L261), with just a few differences.  In particular, note that the goal here is just to verify that the shuffle option is being passed in to the pick_first policy; we don't need to actually verify the behavior of that option, since we're already doing that in the unit test.  So I think that all we really need to do is to check that pick_first chooses a backend randomly each time we send it an update.

Also, note that we don't need multiple CDS resource names, nor do we need multiple EDS resource names.  We can update the default CDS resource just once at the start of the test to configure it to use pick_first, and then we don't need to modify it again.  And we can populate the default EDS resource with the set of backends we want, and then we can update that EDS resource whenever we want -- updating the EDS resource will send an updated address list to the LB policy, even if the new list is identical to the old list.

I suggest something like the following:

```
CreateAndStartBackends(3);
auto eds_resource = BuildEdsResource(EdsResourceArgs(
    {{"locality0", CreateEndpointsForBackends()}}));
// Change cluster to use pick_first with shuffle option.
auto cluster = default_cluster_;
PickFirst pick_first;
pick_first.set_shuffle_address_list(true);
cluster.mutable_load_balancing_policy()
    ->add_policies()
    ->mutable_typed_extension_config()
    ->mutable_typed_config()
    ->PackFrom(pick_first);
balancer_->ads_service()->SetCdsResource(cluster);
// Give 10 tries to get a different backend.
size_t last_seen_backend = backends_.size();
bool saw_different_backend = false;
for (size_t i = 0; i < 10; ++i) {
  // Update EDS resource.  This will send a new address list update to the LB policy.
  balancer_->ads_service()->SetEdsResource(eds_resource);
  // Send an RPC and figure out which backend it went to.
  CheckRpcSendOk(DEBUG_LOCATION);
  size_t backend_index = 0;
  for (; backend_index < backends_.size(); ++backend_index) {
    if (backends_[backend_index]->backend_service()->request_count() > 0) {
      break;
    }
  }
  ASSERT_LT(backend_index, backends_.size());
  // If we found a different backend than last time, then the test succeeded.
  if (last_seen_backend < backends_.size() &&
      backend_index != last_seen_backend) {
    saw_different_backend = true;
    break;
  }
  last_seen_backend = backend_index;
  // Restart the backend to force the client to go IDLE before the
  // next iteration.
  ShutdownBackend(backend_index);
  EXPECT_TRUE(channel_->WaitForStateChange(
      GRPC_CHANNEL_READY, grpc_timeout_seconds_to_deadline(5)));
  EXPECT_THAT(channel_->GetState(false), GRPC_CHANNEL_IDLE);
  StartBackend(backend_index);
}
EXPECT_TRUE(saw_different_backend);
```

Reply:
I arrived to something halfway between my previous version and your proposal. I am toggling between endpoints to ensure EDS update was propagated. Also this assertion - `EXPECT_THAT(channel_->GetState(false), GRPC_CHANNEL_IDLE);` was always failing as it would go to READY. The code seems to work without this now.

Reply:
This approach looks good.

The channel won't go IDLE in this approach, because there's no point at which one of the existing connections fails.  We simply switch from one connection to another due to the address update.